### DISCUSSION TITLE
Why don't we have a faction mode?

### DISCUSSION THREAD

**Discussion by I-am-Erk** [CleverRaven/Cataclysm-DDA#51784] [Ideas]
This is an issue/project I am in the process of writing. Thought folks might enjoy reading it early.

# Problem

One of the most enticing and still-lacking features in this game is NPC interaction and AI with the ability to form your own player faction. This content is hard to get to, and so hard to develop and playtest as it's not attracting players at the rate we'd like.

# Suggestion

I don't think it would actually require very much to turn our existing faction camp system into a rimworld/dwarf fortress style "faction" playmode. This mode should use content from "survivor" mode, but remove the player avatar and instead make you an eye-in-the-sky in a static reality bubble, controlling a group of survivors building a camp.

### First step: UI
As a test of concept what you'd need is a game that loads up in, say, an evac shelter, with just a reality bubble and a few survivors that are part of a faction. You'd be able to select them and view their inventory and @ menu, but not control them. You'd need to be able to assign building zones, construction jobs, and camp missions as in the single player game but through a cursor and direct commands rather than interactions with world objects.

Bizarrely enough, while that wouldn't create a complete game, it should be enough to have an actual functioning mode. There's a lot of content missing and those missions need better balancing, but with those features alone it should be possible to try to play, and the biggest gaps there are with content. If those missions suck, we need to fix them. If there are not enough base expansions, we need to add them.

### Next step: Job queues
We'd also need a queue of jobs waiting to be done, so that survivors without a job could take one from the list. 

We would also want some way to store a job queue for survivors, so Survivor 1 remembers you told them to do four tasks and goes about those tasks in order.  This would itself require you to be able to manually assign a queue of jobs to a survivor when you have them selected, and probably also require some UI windows showing who has what queued.

Job queues also need error handling for when a job gets canceled and must be resumed. For example, if a survivor spots a zombie they should stop what they're doing and deal with it. This means there needs to be a way for urgent jobs, like dealing with a zombie or rescuing a hurt friend, to supplant other jobs in the queue, with at least a somewhat intelligent way of resuming the cancelled job.

### Conflict
We would want some form of conflict. Your base needs to get attacked. Ultimately this should use the wandering hordes mechanics, but since this mode can't function without attacks, we may need some more hard coded things as well.
An immediate option presents itself with scavenging. We should have the option to trigger a 'base attack' on delay, which ideally (for future compatibility) should still use horde mechanics. Basically if you send a survivor on a mission, there should be a chance they attract attention, and a horde spawns to wander back towards your base. 
This could also be done without hordes, just by spawning random monster groups as an effect, but that would mean we'd eventually have to gut that system and make a new one, so it'd be better to start out with hordes. That said, we might also want hte ability to encode specific conflicts as a result of scavenging missions etc. so that we can have other unqiue things, like drawing a jabberwock back to your base.

### Overmap scounting and travel
We would need the ability to send out scouting missions for NPCs to explore the world map for you, since many missions rely on targeting things on the overmap.

We would need an ability for NPCs to drag and use vehicles, so that they can ride bikes and drive cars and fill them with stuff. With new autodrive function this should not be too difficult. at first we'd just use vehicles within the faction camp, I suspect. Once we have that, we should also be able to make scouting missions use vehicles. The next step would be scavenging using vehicles for transport… I suspect that will be more difficult though, as it would require intelligent use of cargo spaces. Perhaps at least some of it could be just hard coded.

### Survivor Needs
As we develop the base's function we should enable things like stamina, weariness, hunger, and sleep for your survivors, so they actually use their beds, and eat the food in the stockpile rather than using calories as a form of currency.

### End of stage 1 notes
As should be apparent, this implementation would be a refined and honed way to force us to approach and solve a lot of the existing problems in NPC AI, by removing the most mitigating factor - the player. Aside from the UI itself, everything here is somethign we want for the survivor game. Isolating it to the faction mechanic alone will bring those problems into focus and also make it way easier to test and improve them.

A lot of stage 1 would hinge on getting content added in systems we already have, like survivor missions and stuff. Stage 2, ironically, would mean removing some of that content and replacing it with emegent behaviours. However, a large chunk of that stopgap stuff will still be necessary (for example, if we design a way to build freeform workshops, we will still want pre-mapped options for them so that later on, we can use them for NPC survivor camps built by AI)

## stage 2: Emergent behaviours
Watch this space...

## Stage 3: advanced needs
This is stuff we don't have yet but have wanted for a long time.
- Survivors should require friendship and confidantes, and develop mental health issues without them - or with them. There should be a role for the faction to have elders and counsellors to deal with the trauma of surviving an apocalypse.
- Sanitation should be an issue you need to deal with, mostly by appointing someone with appropriate skills and knowledge to handle it. It's less that you should have to place the latrines away from clean water and more that you need a designated survivor who, presumably, makes sure that doesn't happen. Sanitation products - both garbage and human waste - should become useable resources in some situations.
- Survivors should have arrays of simple preferences which assign weights to various jobs, foods, clothing, equipment choices, etc. as well as a few invisible stats (some of which we already have) that affect things like whether they'll scream, attack, or sound the alarm when they see a zombie.

**Comment by Qrox** [comment]
> You'd be able to select them and view their inventory and @ menu, but not control them

We should allow assuming direct control of a character in case the AI goes haywire, for example when a character decides to run into a horde of ~goblins~ zombies to get some ~silk~ socks.

**Comment by PatrikLundell** [comment]
Apart from obvious issues, such as the reality bubble and (absent) AI, you also have time as an issue. There's a reason Fortress Mode in Dwarf Fortress operates on an accelerated time while Adventure Mode uses a "normal" time progression. Changing from turn based to RTS is also going to result in issues. Waiting while "dorfs" are working on two week construction projects or mere 2 hour crafting ones in "real" time isn't going to be enjoyable. If you accelerate it you need some safeguards to ensure you're reacting to threats (especially if you deliberately generate threats to attack the base).

I'd say CDDA is quite far from having the beginning of what's needed in both the reality bubble and AI departments. Currently the "no NPC needs" mod is added as the default for a reason. In addition to that, the punishing overheating (and, presumably winter freezing) mechanics is going to be completely overwhelming unless the "dorfs" have the smarts to adjust their clothing to the temperature (which can't be done if they're also going to wear protection) and scavenging would have to somehow have the smarts to bring back what's needed.
For this mode to work, you'd have to have the scavenging parties deplete the resources of the locations they loot, and you'd want to be able to organize fights against the zombies in those locations in a fashion that's at least marginally organized and controlled (Dwarf Fortress' suicide rush is a good example of how not to do it).

Thus, I'd spend a number of years on dedicated work to improve the existing game in the areas required before returning back to thoughts of an RTS mode.
- Make "dorfs" able to feed and clothe themselves without supervision.
- Make them able to follow combat orders (including hit and RUN [as they currently don't know how to run]).
- Make them able to move in formation and act as a formation using formation tactics.
- Implement multiple reality bubbles so it's possible to see what's happening both to scavenging parties and a the base(s).
- Companion driving has been mentioned, and doesn't seem to be out of reach. However, it should be done in their own reality bubbles.
- Improve current mode base missions to a level where they would make sense in an RTS mode. This would mean sending the companions out in their own reality bubbles and actually perform the tasks they're sent to perform. This, in turn, would require them to react appropriately to threats.
- You'd probably need larger reality bubbles as well (or assign fixed ones to the "expansion slots" of the base) to avoid hordes spawning inside the base just because the reality bubble happens to be centered on the other end of the base. Even with a reality bubble fixed at the center of the base, enemies would spawn just next to it with the current bubble size.

**Comment by 25yearsMeanwhile** [comment]
Why does there need to be a reality bubble? Why not faction management within the regular game, a la Kenshi?

**Reply by PatrikLundell** [reply]
The reality bubble is an artifact created to reduce the amount of calculation needed to a level that isn't unreasonably high (most of the time). Everything outside of the bubble is either frozen completely and sometimes fast-forwarded when entering the bubble and sometimes not, or subjected to simplified calculations.

It may or may not be possible to modify the game to support multiple reality bubbles, but that would probably be a rather hefty change.
